File: projecteuler/euler011.rb
grid = Array.new(20)
20.times { |i| grid[i] = gets.strip.split.map(&:to_i) }

max = -1

2.upto(18) do |i|
  2.upto(18) do |j|
    -2.upto(1) do |i1|
      hproduct, vproduct = 1, 1
      -2.upto(1) { |j1| hproduct *= grid[i + i1][j + j1] }
      -2.upto(1) { |j1| vproduct *= grid[j + j1][i + i1] }
      max = [hproduct, vproduct, max].max
    end
    rd = grid[i - 2][j - 2] * grid[i - 1][j - 1] * grid[i][j] * grid[i + 1][j + 1]
    ld = grid[i - 2][j + 1] * grid[i - 1][j] * grid[i][j - 1] * grid[i + 1][j - 2]
    max = [max, rd, ld].max
    j += 1
  end
  i += 1
end

puts max

File: projecteuler/Euler011.java
import java.io.*;
import java.util.*;
import java.text.*;
import java.math.*;
import java.util.regex.*;

public class Solution {
  public static void main(String[] args) {
    Scanner scanner = new Scanner(System.in);
    int[][] grid = new int[20][20];
    int i, j, l, r, max = -1, hproduct, vproduct, rd, ld;

    for (i = 0; i < 20; i++) {
      for (j = 0; j < 20; j++) {
        grid[i][j] = scanner.nextInt();
      }
    }

    for (i = 2; i <= 18; i++) {
      for (j = 2; j <= 18; j++) {
        for(l = -2; l <= 1; l++) {
          hproduct = 1;
          vproduct = 1;
          for(r = -2; r <= 1; r++) {
            hproduct *= grid[i + l][j + r];
          }
          for(r = -2; r <= 1; r++) {
            vproduct *= grid[j + r][i + l];
          }
          if (hproduct > vproduct && hproduct > max) {
            max = hproduct;
          }
          if (vproduct > hproduct && vproduct > max) {
            max = vproduct;
          }
        }
        rd = grid[i - 2][j - 2] * grid[i - 1][j - 1] * grid[i][j] * grid[i + 1][j + 1];
        ld = grid[i - 2][j + 1] * grid[i - 1][j] * grid[i][j - 1] * grid[i + 1][j - 2];
        if (rd > ld && rd > max) {
          max = rd;
        }
        if (ld > rd && ld > max) {
          max = ld;
        }
      }
    }
    System.out.println(max);
  }
}
